keywrap/pkcs11/keywrapper_pkcs11.go

```diff
@@ -0,0 +1,233 @@
+)
+
+var (
+	KeyLabel   = "imgenc"
+	OAEPParams = &pkcs11.OAEPParams{
+		HashAlg:    pkcs11.CKM_SHA_1,
+		MGF:        pkcs11.CKG_MGF1_SHA1,
+		SourceType: pkcs11.CKZ_DATA_SPECIFIED,
+		SourceData: nil,
+	}
+)
+
+type pkcs11KeyWrapper struct {
+}
+
+func NewKeyWrapper() keywrap.KeyWrapper {
+	return &pkcs11KeyWrapper{}
+}
+
+func (kw pkcs11KeyWrapper) GetAnnotationID() string {
+	return "org.opencontainers.image.enc.keys.experimental.pkcs11"
+}
+
+func (kw pkcs11KeyWrapper) WrapKeys(ec *config.EncryptConfig, optsData []byte) ([]byte, error) {
+	if len(ec.Parameters["modules"]) == 0 || len(ec.Parameters["pins"]) == 0 {
+		return nil, nil
+	}
+	p11ctx, session, err := loginDevice(ec.Parameters["modules"], ec.Parameters["pins"])
+	defer closeModule(p11ctx, session)
+	// no recipients
+	if err != nil {
+		return nil, err
+	}
+
+	pub, err := getPublickey(KeyLabel, p11ctx, session)
+	if err != nil {
+		return nil, err
+	}
+
+	err = p11ctx.EncryptInit(session, []*pkcs11.Mechanism{pkcs11.NewMechanism(pkcs11.CKM_RSA_PKCS_OAEP, OAEPParams)}, pub)
+	if err != nil {
+		return nil, errors.Wrap(err, "Module EncryptInit error")
+	}
+
+	encrypt, err := p11ctx.Encrypt(session, optsData)
+	if err != nil {
+		return nil, errors.Wrap(err, "Encrypt with Module failed")
+	}
+
+	return encrypt, nil
+}
+
+func (kw pkcs11KeyWrapper) UnwrapKey(dc *config.DecryptConfig, encrypted []byte) (plain []byte, err error) {
+	p11ctx, session, err := loginDevice(dc.Parameters["modules"], dc.Parameters["pins"])
+	defer closeModule(p11ctx, session)
+
+	priv, err := findObject(p11ctx, session, pkcs11.CKO_PRIVATE_KEY, KeyLabel)
+	if err != nil {
+		return nil, err
+	}
+
+	err = p11ctx.DecryptInit(session, []*pkcs11.Mechanism{pkcs11.NewMechanism(pkcs11.CKM_RSA_PKCS_OAEP, OAEPParams)}, priv)
+	if err != nil {
+		return nil, errors.Wrap(err, "Decrypt init failed")
+	}
+	plain, err = p11ctx.Decrypt(session, encrypted)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "Decrypt failed")
+	}
+
+	return
+}
+
+func (kw pkcs11KeyWrapper) NoPossibleKeys(dcparameters map[string][][]byte) bool {
+	return len(kw.GetPrivateKeys(dcparameters)) == 0
+}
+
+// GetPrivateKeys return private key handle
+func (kw pkcs11KeyWrapper) GetPrivateKeys(dcparameters map[string][][]byte) [][]byte {
+	return dcparameters["modules"]
+}
+
+// GetKeyIdsFromWrappedKeys converts the base64 encoded Packet to uint64 keyIds;
+// We cannot do this with pkcs11
+func (kw pkcs11KeyWrapper) GetKeyIdsFromPacket(packet string) ([]uint64, error) {
+	return nil, nil
+}
+
+// GetRecipients converts the wrappedKeys to an array of recipients
+// We cannot do this with pkcs11
+func (kw pkcs11KeyWrapper) GetRecipients(packet string) ([]string, error) {
+	return []string{"[pkcs11]"}, nil
+}
+
+// getPublickey get module public key, generate a rsa key if it doesn't exist
+// TODO: experimental, use RSA Key for wrap/unwrap
+func getPublickey(label string, p *pkcs11.Ctx, sh pkcs11.SessionHandle) (pub pkcs11.ObjectHandle, err error) {
```

Comment:
Can we rename this to `getPublicKeyOrCreate`?

Reply:
fixed.